core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestSimplifyExpressions.java

```diff
@@ -54,54 +57,93 @@
+        assertSimplifies("NOT (X OR Y OR (Z OR V))", "(NOT X) AND (NOT Y) AND ((NOT Z) AND (NOT V))", BOOLEAN);
+
+        assertSimplifies("NOT (X IS DISTINCT FROM Y)", "NOT (X IS DISTINCT FROM Y)", BOOLEAN);
+    }
+
+    @Test
+    public void testLikeExpressions()
+    {
+        assertSimplifies("name LIKE '%'", "name IS NOT NULL", createCharType(2));
+        assertSimplifies("name LIKE '%%'", "name IS NOT NULL", createCharType(2));
+        assertSimplifies("name LIKE '%%%%'", "name IS NOT NULL", createCharType(10));
+        assertSimplifies("name LIKE '%%%%' ESCAPE '\\'", "name IS NOT NULL", createCharType(10));
+        assertSimplifies("name LIKE '中文%abc字母😂'", "name LIKE '中文%abc字母😂'", createCharType(10));
+        assertSimplifies("name LIKE '中文%abc字母😂' ESCAPE '\\'", "name LIKE '中文%abc字母😂' ESCAPE '\\'", createCharType(10));
+
+        assertSimplifies("name LIKE '%'", "name IS NOT NULL", createVarcharType(2));
+        assertSimplifies("name LIKE '%%'", "name IS NOT NULL", createVarcharType(2));
+        assertSimplifies("name LIKE '%%%%'", "name IS NOT NULL", createVarcharType(10));
+        assertSimplifies("name LIKE '%%%%' ESCAPE '\\'", "name IS NOT NULL", createVarcharType(10));
+        assertSimplifies("name LIKE '中文%abc字母😂'", "name LIKE '中文%abc字母😂'", createVarcharType(10));
+        assertSimplifies("name LIKE '中文%abc字母😂' ESCAPE '\\'", "name LIKE '中文%abc字母😂' ESCAPE '\\'", createVarcharType(10));
+
+        assertSimplifies("name LIKE '%'", "name IS NOT NULL", VARCHAR);
+        assertSimplifies("name LIKE '%%'", "name IS NOT NULL", VARCHAR);
+        assertSimplifies("name LIKE '%%%%'", "name IS NOT NULL", VARCHAR);
+        assertSimplifies("name LIKE '%%%%' ESCAPE '\\'", "name IS NOT NULL", VARCHAR);
+        assertSimplifies("name LIKE '中文%abc字母😂'", "name LIKE '中文%abc字母😂'", VARCHAR);
+        assertSimplifies("name LIKE '中文%abc字母😂' ESCAPE '\\'", "name LIKE '中文%abc字母😂' ESCAPE '\\'", VARCHAR);
+
+        // test with the like constant
+        assertSimplifies("name LIKE 'This is a constant'", "name = VARCHAR 'This is a constant'", VARCHAR);
+        assertSimplifies("name LIKE '!@#$#!'", "name = VARCHAR '!@#$#!'", VARCHAR);
+        assertSimplifies("name LIKE '中文abc字母😂'", "name = VARCHAR '中文abc字母😂'", VARCHAR);
+
+        // test with the escape char
+        assertSimplifies("name LIKE '\\%' ESCAPE '\\'", "name = VARCHAR '%'", VARCHAR);
+        assertSimplifies("name LIKE 'abc\\%' ESCAPE '\\'", "name = VARCHAR 'abc%'", VARCHAR);
+        assertSimplifies("name LIKE '\\%%%%' ESCAPE '\\'", "name LIKE '\\%%%%' ESCAPE '\\'", VARCHAR);
+        assertSimplifies("name LIKE '%\\%\\%%%%' ESCAPE '\\'", "name LIKE '%\\%\\%%%%' ESCAPE '\\'", VARCHAR);
+        assertSimplifies("name LIKE '%%' ESCAPE '%'", "name = VARCHAR '%'", VARCHAR);
+        assertSimplifies("name LIKE '%%%%' ESCAPE '%'", "name = VARCHAR '%%'", VARCHAR);
+        assertSimplifies("name LIKE '中文%%abc字母😂' ESCAPE '%'", "name = VARCHAR '中文%abc字母😂'", VARCHAR);
```

Comment:
These should all be boolean expressions, so using VARCHAR as a type is incorrect. Why do the tests not fail?

Reply:
The type is used for describing the symbol reference not the whole expression, here is the `name`. I removed the type param for the tests that doesn't have symbol reference, e,g:`CAST(12300000000 AS varchar(11))`, and added a `assertSimplifies` method without type param for it.

Reply:
Make sense, how about put the symbol types in the test cases? Like `assertSimplifies("X AND Y", "X AND Y", ImmutableMap.of("X", BOOLEAN, "Y", BOOLEAN));`

Reply:
Yes, that would work.